src/v/pandaproxy/sharded_client_cache.cc

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "pandaproxy/sharded_client_cache.h"
+
+#include "ssx/future-util.h"
+
+using namespace std::chrono_literals;
+
+namespace pandaproxy {
+
+static constexpr auto clean_timer_period = 10s;
+
+ss::future<> sharded_client_cache::start(
+  ss::smp_service_group sg,
+  YAML::Node const& cfg,
+  config::binding<size_t> max_size,
+  config::binding<std::chrono::milliseconds> keep_alive) {
+    _smp_opts = ss::smp_submit_to_options{sg};
+
+    _clean_timer.set_callback([this] {
+        ssx::spawn_with_gate(_gate, [this] {
+            return _cache.invoke_on_all(
+              _smp_opts,
+              [](kafka_client_cache& cache) { cache.clean_stale_clients(); });
+        });
+    });
+
+    _clean_timer.arm_periodic(clean_timer_period);
+
+    // Pass configs by value here or else risk passing the binding
+    // across shards (i.e., fails oncore_debug_verify)
+    return _cache.start(cfg, max_size(), std::move(keep_alive()));
+}
+
+ss::future<> sharded_client_cache::stop() {
+    // Reset callback first because the original
+    // captured "this" and cancelling on that could
+    // be unsafe.
+    _clean_timer.set_callback([] {});
```

Comment:
clearing the callback should be unnecessary. even though the lambda holds a reference to this, the timer and its lambda and *this effectively all have the same life time. all you need to is cancel it.

Reply:
I'm getting mixed reviews on this. I'll likely keep this in depending on your response to https://github.com/redpanda-data/redpanda/pull/6452#discussion_r978654668

Reply:
I had this is other parts of the code, because the callback held a `shared_from_this()`, preventing it from ever being destroyed. I don't think it's needed here.

Reply:
> shared_from_this

oh, like a self-referential situation in which the lambda was keeping *this alive? i could see something like that. but let's try to avoid relying on needing to clear the timer lambda.

Ben, do you have a reference to where you needed to do this? It seems like something that can be worked around.